File: src/xSTIR/pSTIR/tests/test2.py
'''OSEM reconstruction test.

Usage:
  test2 [--help | options]

Options:
  -r, --record   record the measurements rather than check them
  -v, --verbose  report each test status
'''

## CCP PETMR Synergistic Image Reconstruction Framework (SIRF)
## Copyright 2015 - 2017 Rutherford Appleton Laboratory STFC
##
## This is software developed for the Collaborative Computational
## Project in Positron Emission Tomography and Magnetic Resonance imaging
## (http://www.ccppetmr.ac.uk/).
##
## Licensed under the Apache License, Version 2.0 (the "License");
##   you may not use this file except in compliance with the License.
##   You may obtain a copy of the License at
##       http://www.apache.org/licenses/LICENSE-2.0
##   Unless required by applicable law or agreed to in writing, software
##   distributed under the License is distributed on an "AS IS" BASIS,
##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##   See the License for the specific language governing permissions and
##   limitations under the License.

__version__ = '0.1.0'
from docopt import docopt
args = docopt(__doc__, version=__version__)

import math

from pSTIR import *

record = args['--record']
verbose = args['--verbose']

def main(verb = False):

    test = pTest('test2.txt', record)
    test.verbose = verb

    msg_red = MessageRedirector()

    data_path = petmr_data_path('pet')
    raw_data_file = existing_filepath(data_path, 'my_forward_projection.hs')
    acq_data = AcquisitionData(raw_data_file)
    test.check(acq_data.norm())

    image = acq_data.create_uniform_image(1.0)
    test.check(image.norm())

    acq_model = AcquisitionModelUsingRayTracingMatrix()
    acq_model.set_up(acq_data, image)

    obj_fun = make_Poisson_loglikelihood(acq_data)
    obj_fun.set_acquisition_model(acq_model)

    num_subsets = 12
    recon = OSMAPOSLReconstructor()
    recon.set_objective_function(obj_fun)
    recon.set_num_subsets(num_subsets)
    recon.set_input(acq_data)
    print('setting up, please wait...')
    recon.set_up(image)

    recon.set_current_estimate(image)

    num_iterations = 2
    for iteration in range(num_iterations):
        print('\n------------- iteration %d' % iteration)
        recon.update_current_estimate()
    test.check(image.norm())

    print('projecting...')
    simulated_data = acq_model.forward(image)
    diff = simulated_data * (acq_data.norm()/simulated_data.norm()) - acq_data
    print('relative residual norm: %e' % (diff.norm()/acq_data.norm()))
    test.check(diff.norm())

    return test.failed, test.ntest

if __name__ == '__main__':

    try:
        failed, ntest = main(verbose)
        if failed == 0:
            if not record:
                print('all tests passed')
            else:
                print('%d measurements recorded' % ntest)
            sys.exit(0)
        else:
            print('%d of the tests failed' % failed)
            sys.exit(failed)

    except error as err:
        # display error information
        print('??? %s' % err.value)
        sys.exit(-1)

File: src/xGadgetron/mGadgetron/tests/test2.m
function [failed, ntests] = test2(record, engine)
% MR test set 2.

% GRAPPA reconstruction and the illustration of the use of 
% Acquisition Model projections.

% CCP PETMR Synergistic Image Reconstruction Framework (SIRF).
% Copyright 2015 - 2017 Rutherford Appleton Laboratory STFC.
% Copyright 2015 - 2017 University College London.
% 
% This is software developed for the Collaborative Computational
% Project in Positron Emission Tomography and Magnetic Resonance imaging
% (http://www.ccppetmr.ac.uk/).
% 
% Licensed under the Apache License, Version 2.0 (the "License");
% you may not use this file except in compliance with the License.
% You may obtain a copy of the License at
% http://www.apache.org/licenses/LICENSE-2.0
% Unless required by applicable law or agreed to in writing, software
% distributed under the License is distributed on an "AS IS" BASIS,
% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
% See the License for the specific language governing permissions and
% limitations under the License.

% Select and import SIRF MATLAB MR package so that SIRF MR objects can be 
% created in this function without using the prefix 'MR.'
if nargin < 2
    engine = [];
end
if nargin < 1
    record = false;
end
% import_str = set_up_MR(engine);
% eval(import_str)
MR = set_up_MR(engine);

test = mUtilities.mTest('test2.txt', record);

filename = 'simulated_MR_2D_cartesian_Grappa2.h5';
pathname = mUtilities.examples_data_path('MR');
acq_data = MR.AcquisitionData(fullfile(pathname, filename));
test.check(acq_data.norm())

prep_gadgets = {'RemoveROOversamplingGadget'};
processed_data = acq_data.process(prep_gadgets);
test.check(processed_data.norm())

recon = MR.CartesianGRAPPAReconstructor();
recon.compute_gfactors(false);
recon.set_input(processed_data);
recon.process();
complex_images = recon.get_output();
test.check(complex_images.norm())

processed_data.sort()
csms = MR.CoilSensitivityData();
csms.calculate(processed_data)

am = MR.AcquisitionModel(processed_data, complex_images);
am.set_coil_sensitivity_maps(csms)
fwd_acqs = am.forward(complex_images);
fwd_acqs_norm = fwd_acqs.norm();
test.check(fwd_acqs_norm)

acqs_diff = fwd_acqs - processed_data;
rr = acqs_diff.norm()/fwd_acqs_norm;
test.check(rr, 1e-4)

bwd_images = am.backward(processed_data);
imgs_diff = bwd_images - complex_images;
rd = imgs_diff.norm()/complex_images.norm();
test.check(rd, 1e-4)

xFy = processed_data * fwd_acqs;
Bxy = bwd_images * complex_images;
test.check(abs(real(xFy)/real(Bxy) - 1.0), 1e-4)

failed = test.failed;
ntests = test.ntest;

if record
    fprintf('%d measurements recorded\n', ntests)
elseif failed == 0
    fprintf('all %d tests passed\n', ntests);
else
    fprintf('%d out of %d tests failed\n', failed, ntests);
end
end